### ISSUE TITLE
Eigensolving TN operators

### ISSUE THREAD

**Issue by Thenerdstation** [google/TensorNetwork#548]
We should build a method to allow users to find the top eigenvector of a linear operator defined by a tensor network.

My API proposal is as such

```
tensor = tn.eigensolution(in_edges=list_of_edges, out_edges=list_of_edges)
```
Where `tensor.shape == [edge.dimension for edge in in_edges]` and `tensor` is the actual `tensor` and NOT a `Node`. 

You can easily find the necessary nodes via `tn.reachable`, and to prevent smashing the given tensornetwork, we can safely copy all of the nodes via `tn.replicate_nodes`.

**Comment by Thenerdstation** [comment]
With the new lanczos algorithm written up, this is actually a very viable function now.

**Comment by mganahl** [comment]
I agree we should make sparse solvers directly accessible for users, not only via the backend object. I see two potential problems with the API above. 

First, linear operators come with different properties (hermitian, symmetric, positive definite, ....), and the choice of which algorithm to use has to take that into account. Instead of having a single function `eigensolution`, I'd rather have different functions like `eigsh_lanczos`, `eigs_arnoldi`, `eigsh_davidson` and so on. The alternative to that would be to pass flags to `eigensolution`, which typically is a major source of obscure bugs, so I'd like to avoid it as much as possible. 

The second point is about passing edges. The problem is that the network behind the passed edges can be arbitrarily complicated. The network contraction needs to be called many times during a single call to `eigensolution`, so to get good performance we need a good contraction path, e.g. from an automatic contraction path finder (they are very costly though). In a typical setting `eigensolution` is called many times, e.g. in a `for` or `while` loop. To maintain performance, we have to reuse the previously computed contraction paths, and I'm not sure if there is a clean and easy way to keep track of all these paths for an arbitrary number of networks. 

I would instead keep the API identical to the one used in the backend, i.e.
```python 
tn.eigs_arnoldi(matvec, args_to_matvec, ...)
```
so that a user can do
```python
#matvec maps an input `vector` to an output vector of same dimension.
def matvec(t1, t2, t3,..., tn, vector):
  """construct the network here 
  """
  return result

eigvals, eigvecs = tn.eigs_arnoldi(matvec,[t1, t2, t3, ..., tn],...)
```

**Comment by Thenerdstation** [comment]
> First, linear operators come with different properties (hermitian, symmetric, positive definite, ....

That's a really good point. Yeah, lets change the api the be `eigensolve_hermitian`, `eigensolve_symmetric`, etc.

> The second point is about passing edges. The problem is that the network behind the passed edges can be arbitrarily complicated....

The problem I see with having a user provide their own `matvec` function is that graphs are mutable, so the second time the matvec function is called, there would be lots of errors. By having the user provide the `edges` instead, we can copy the graph via `tn.reachable(edges)` and `tn.copy`. Then we can safely build our own `matvec` function that preservers the original network.

So how about this instead?

```tn.eigensolve_hermitian(input_edges, output_edges, nodes, contraction_algorithm, *args)`.

Here, the `input_edges` and `output_edges` remain the same, but we add two additional parameters `nodes`, and `contraction_algorithm. If a user wants to hand define a contraction path, then they would pass a `contraction_algorithm` that's defined like:

```python
def my_path(nodes):
  # The "input vector" will always be appended to the start of the list.
  tmp = nodes[0] @ nodes[1]
  return tmp @ nodes[2]

tn.eigensolution_hermitian(inputs, outputs, my_nodes, my_path)
```

Then we can take their function and wrapped it safely into our own `matvec` function. Additionally, a user could freely pass any of the `tn.contractors`. If no method is passed, we'll default to using `auto`.

**Comment by Thenerdstation** [comment]
`nodes` would also be an optional argument. If `nodes` is not passed, we'll default to whatever `list(tn.reachable(input_edges + output_edges))` returns

**Comment by gevenbly** [comment]
Hey guys, I am not so familiar with the details and workings of the tensornetwork library yet, but I thought I could offer some general suggestions.

In practice, one often needs to solve eigenvectors of a linear operator defined by a weighted sum of several different networks (or maybe things even more complicated), so I think that it is essential to allow the user to define their own `matvec`  function.

As for the API, I would suggest sticking with (or adopting something close to) the existing one for sparse eigensolvers: `scipy.sparse.linalg.eigsh` and `scipy.sparse.linalg.eigs`, which can be made compatible with an arbitrary `matvec` function by using `scipy.sparse.linalg.LinearOperator` or similar. The `eigsh` and `eigs` already employ the optimal sparse eigensolver (given the symmetry of the linear operator) and accept a host of optional arguments `(k, M, sigma, which, v0, ncv, maxiter, tol, return_eigenvectors)` which I feel should be supported (in my own work, at least, I routinely require the use of all of these).

**Comment by Thenerdstation** [comment]
Hm, in that case it might be nice to actually just have these in JAX proper. Arbitrary `matvec` functions seems somewhat out of scope for this library, where as putting `eigsh` into JAX would be helpful for many more users.

**Comment by mganahl** [comment]
It would indeed be, as would `eigs`. 

I actually think that arbitrary `matvec` has to absolutely be within the scope of this library. Sparse solvers are needed in so many tensor network algorithms that not having them would be an absolute no-go (I can say with confidence that I would just not use the library in that case). As @gevenbly has pointed out, not all `matvec`s can be easily written as a simple tensor network, and we have to support such cases as well. I think the best way of providing this is via user defined `matvecs` as in the above-illustrated way.

One of the main points of tensornetwork is to provide functionality to write backend independent tensor network algorithms, so we have to provide `eigs` and `eigsh` not only in the backend but also via the front end. If we don't provide the `matvec` fuctionality for `eigs` and `eigsh` to the user via the front end, I'm concerned that many will just opt to not use it.

I think to maintain backend independence it makes sense to have `eigsh` take `Node` objects and return `Node` objects.

The good news is that we essentially have almost all the pieces. For user-defined contraction paths this would work fine:
```python
def matvec([node1, node2, ...], vector_node,...):
  #user uses whatever API she prefers:
  #tensornetwork API
  node1[0] ^ node2[5]
  ...
  res = ... #the final node
  #or ncon
  res = tn.ncon([node1, node2, ...], [[...],[...],....])
  return res

eigvals, eignodes = eigsh(matvec, [n1,n2,n3],...) #returns eigenvalues and eigenvectors, wrapped into `Node` objects.
```
Optimal contraction paths could be implemeted using e.g. `ncon`, or have `matvec` take the path as argument. In `pytorch` and `numpy` all of this is straightforward and already supported. 

JAX is a bit more tricky because we need to jit the code, and passing nodes or other types as arguments to jitted functions is by default not permitted, but we can register user-defined types with JAX, which makes them passable. I did some quick tests and this seems to work with `Node` objects, so the above API would likely work there as well. The only remaining question mark is `tensorflow`, as I am not sure if it is possible to pass user-defined arguments to `tf.fuction` wrapped functions. I'm confident though we can find a solution there as well to get a seamless API for all backends. 

I'm open to using an API with a `LinearOperator` class, or just passing plain functions. One thing we probably need to keep though is passing an `args` argument containing the input arguments to the `matvec` operation. JAX strongly recommends jitted functions being closures, otherwise unexpected side effects could appear. Unless I fully understand the implications of passing non-closures I'd rather stick with their recommendation. This essentially would force this part of the API on us.

**Comment by mganahl** [comment]
In fact, there are even easier workarounds than using the JAX pytree, and those also work with tensorflow

**Comment by Thenerdstation** [comment]
Let's just completely ignore tensorflow for this use case, as I'm not convinced scientists will ever use the TF backend, especially since we highly encourage using JAX when you want to use hardware accelerators anyway. TF and PyTorch are mainly supported for ML use cases, which likely would never need `eigs`. 

> I think to maintain backend independence it makes sense to have eigsh take Node objects and return Node objects.

I'm not sure that's the best idea. The first problem is what do you do when these nodes are already connected to each other. Raise an error? Assume it's ok? It's far too much of a footgun for users. Like you said, JAX expectes functions to be closures, and the tn.Nodes are highly mutable in our current implementation. Another footgun.

One option to work around this is to instead pass tensors instead of nodes to the `matvec` function. But if we do that, then when is the user using a tensor network? It feels like its at the wrong level of abstraction.

**Comment by mganahl** [comment]
> Let's just completely ignore tensorflow for this use case, as I'm not convinced scientists will ever use the TF backend, especially since we highly encourage using JAX when you want to use hardware accelerators anyway. TF and PyTorch are mainly supported for ML use cases, which likely would never need `eigs`.
> 
> > I think to maintain backend independence it makes sense to have eigsh take Node objects and return Node objects.
> 
> I'm not sure that's the best idea. The first problem is what do you do when these nodes are already connected to each other. Raise an error? Assume it's ok? It's far too much of a footgun for users. Like you said, JAX expectes functions to be closures, and the tn.Nodes are highly mutable in our current implementation. Another footgun.

This should clearly raise an error, since all previously defined connections will be erased during the first call to `matvec`.

If we require input nodes to be unconnected I don't see a problem with the closures. In that case, `Nodes` can essentially be regarded as equivalent to jax-arrays. Unless the user has decided to do some really shady things with references to `Node.tensor`, I don't see how this could break.  
> 
> One option to work around this is to instead pass tensors instead of nodes to the `matvec` function. 

I think this would be possible too.

>But if we do that, then when is the user using a tensor network? It feels like its at the wrong level of abstraction.

He uses it exactly where it has always been used: inside the matvec function. I think we should not regard a tensor network as a single entity that is to be defined at one place in the code at the beginning and maintains its connections for eternity. If this is our design philosophy, that would lead to a very brittle API. In practice, every tensor network calculation is broken down into contractions and decompositions over small subsets of tensors. There may be special cases of tensor networks where a global approach could be feasible and convenient, and they are worth exploring. But for the large majority, it will lead to an overly verbose and complicated code. 

The small subset of tensors is the entity we should focus on, and for which we should provide backend-independent code to do contractions and dense and sparse decomposition.

**Comment by Thenerdstation** [comment]
Alright, you've convinced me.

Let's do the `matvec` functions with tensors being passed in. 

Let's make the API look like this?

```python
def eigs(matvec, tensors, ... <eigs_args>...):
  ...
```

And then the `matvec` functions should follow the API spec of
```python
def matvec(vector, tensors):
  ...
```

My annoyance with this API is that `matvec` has to be define with tensors, even if they don't use them.

**Comment by mganahl** [comment]
Sounds good. The current backends expect `matvec(tensors, vector)`, but I'll fix this to conform to the above. 

I agree,  having `Node`s would be the natural choice. This is a matter of making this work with JAX, I'll have a look at this.

**Comment by mganahl** [comment]
Is there a reason for not using 
```python 
def matvec(vector, n1, n2, ,...):
  #do stuff
```
and call `matvec(vector, *args)` inside eigsh?

**Comment by gevenbly** [comment]
The format you suggest:
`def matvec(vector, n1, n2, ,...):`
is the common one I have seen before (e.g. also inbuilt with MATLAB eigs and whatnot).

In general the arguments `n1,n2,...` could be of any type, not just a set of tensors. It is common for `matvec` functions to need other scalar parameters, or to accept custom objects (e.g. for specifying contraction order)

**Comment by Thenerdstation** [comment]
Yes `def f(matvec, *args)` sounds good to me too.

**Comment by mganahl** [comment]
@gevenbly that could become very tricky. The problem is that this will likely lead to issues with JAX. We are currently wrapping `matvec` into `jax.tree_util.Partial` so that it can be passed to a jitted eigsh solver. We need to jit the eigensolver to get good performance, and we only want to jit it once (or a very small number of times). 
Currently, we wrap `matvec` into `tree_util.Partial` (another necessary technicality) and pass it to a jitted jax-lanczos solver.  Upon jitting, the arguments to `matvec` get traced and can't be used any more in the originally inteded ways (essentially, the code crashes). Not sure how to prevent this. @Thenerdstation do you know how to let python objects pass into `matvec` in that case? using `static_argnums` on matvec seems not to work in that case.

**Comment by Thenerdstation** [comment]
Well my advice was going to be to use `static_argnums`. Why doesn't that work currently? I'm still confused as to why you need `Partial` instead of just normal `jit`.

**Comment by Thenerdstation** [comment]
Another option is to just tell users to be more functional. There's no reason `matvec` can't be the result of a partial.

```python
def my_matvec(vector, tensors, some_special_thing):
  pass

tmp_f = partial(my_matvec, some_special_thing="the thing")
tn.eigs(tmp_f)
```
Jax has made me love this functional approach, so I wanna stick with it.

**Comment by Thenerdstation** [comment]
Doing it that way also makes it more clear to the user when the function is getting recompiled. It's just whenever you give it a new function definition.

**Comment by mganahl** [comment]
I actually like that approach

**Comment by mganahl** [comment]
Looks to be a very clean and easy solution. So we make it very clear in the doc that the `matvec` passed to `eigsh` can only accept tensor objects, and give a small example how partials can be used to accomplish more complex behaviour (with loud warnings about recompilation for the case of JAX).

**Comment by Thenerdstation** [comment]
Yes, these all sound good to me.

**Comment by mganahl** [comment]
With a bit of internal juggling we can have `eigsh` accept `matvec`s that take and return `Node` objects. 
```python 
cachedfuns={}
def eigsh(mv, nodes, initial_node):
    be=nodes[0].backend    
    if mv not in cachedfuns:
        def _mv(tensor, *args, backend):
            _nodes = [tn.Node(arg, backend=backend) for arg in args]
            _node = tn.Node(tensor, backend=backend)
            res = mv(_node, *_nodes)
            return res.tensor
        cachedfuns[mv] = partial(_mv, backend=be)
    
    eta, U = backend.eigsh_lanczos(cachedfuns[mv],[n.tensor for n in nodes],initial_node.tensor)
    return [tn.Node(e,backend=backend) for e in eta],[tn.Node(u,backend=backend) for u in U]
```

**Comment by mganahl** [comment]
function caching again to avoid JAX overhead

**Comment by Thenerdstation** [comment]
I mean, do we really need it to return `Nodes`?

**Comment by mganahl** [comment]
no, tensors would also be fine. I think right now the API is a bit ambiguous as to when 
using `Node` vs when using tensors. A user can use`Node` as containers for the different tensor types throughout the whole simulation. Code written this way would be automatically backend independent. Or the user can use tensor objects, and wrap them into `Node` whenever he needs to, and extract tensors back afterwards. It depends on the users preference.

**Comment by Thenerdstation** [comment]
Lets just stick with tensors for now. If there is a strong want for passing in `Nodes` we'll add it in as it doesn't seem too hard to do.

**Comment by mganahl** [comment]
👍
